### DISCUSSION TITLE
How do I bind button commands in the ItemControl template

### DISCUSSION THREAD

**Discussion by coinsoundsbetter** [AvaloniaUI/Avalonia#14189] [Q&A] [answered]
As in the following example, to display multiple Buttons dynamically, I define an ItemControls and bind the FuncShowButtons collection in the FuncShowViewModel to an ItemSource. My intention is to be able to respond to a Button event in the FuncShowViewModel when I click on a dynamically displayed button -- but what should I do when I work with the layout code as follows and the button displays in an unclickable state with no compilation error?

So everyone can see more clearly, In the layout code below, I annotate the desired data source for the data with ():

`<StackPanel VerticalAlignment="Center" HorizontalAlignment="Center">
        <ItemsControl ItemsSource="{Binding Path=FuncShowButtons(List<FuncShowButtonModel>)}">
            <ItemsControl.ItemTemplate>
                <DataTemplate>
                    <Button Content="{Binding Name(FuncShowButtonModel)}"
                            Command="{Binding Path=SelectFuncCommand(FuncShowViewModel), RelativeSource={RelativeSource FindAncestor, AncestorType=viewModels:FuncShowViewModel}}">
                    </Button>
                </DataTemplate>
            </ItemsControl.ItemTemplate>
        </ItemsControl>
    </StackPanel>`

**Comment by thevortexcloud** [comment, marked as answer]
> , AncestorType=viewModels:FuncShowViewModel}

The ancestor will never be a view model. It will be a control. You can instead bind to the DataContext of the control which will likely be a view model.

**Reply by coinsoundsbetter** [reply]
Thank you for your answer! I tried to make this change, but it shows **Error AVLN:0004 Avalonia: Unable to resolve property or method of name 'SelectFuncCommand' on type 'System.Object'**, I looked at a lot of sources, and they all looked like this, I can't understand why it went wrong. Can you help me point out my mistakes?
This is my modification part:
From:
`<DataTemplate>
                    <Button Content="{Binding Name(FuncShowButtonModel)}"
                            Command="{Binding Path=SelectFuncCommand(FuncShowViewModel), RelativeSource={RelativeSource FindAncestor, AncestorType=viewModels:FuncShowViewModel}}">
                    </Button>
                </DataTemplate>`
To:
`<UserControl.DataContext>
        <viewModels:FuncShowViewModel/>
    </UserControl.DataContext>`
+
`<DataTemplate>
                    <Button Command="{Binding  RelativeSource={RelativeSource AncestorType={x:Type UserControl}, Mode=FindAncestor}, Path=DataContext.SelectFuncCommand}"></Button>
                </DataTemplate>`

**Reply by thevortexcloud** [reply]
You need to cast the datacontext or use a reflection binding. Something like `Path=((FuncShowViewModel)DataContext).SelectFuncCommand`

**Reply by coinsoundsbetter** [reply]
Wow, now it seems that it is indeed missing a type conversion, thank you for your help!

**Reply by thevortexcloud** [reply]
You are welcome. This is because the type of `DataContext` is always `object`. You however wont have this problem if you use a reflection binding or disable compiled bindings. You may have other issues though if you decide to do that. But sometimes it's the easiest way.

**Reply by coinsoundsbetter** [reply]
So in such dynamically generated interactive buttons, the more common practice is "reflection binding"? In my gut, should we try to avoid "reflective binding"? (Because it causes us to fail to detect potential risks during compilation)
But in the official MusicStore example, I've seen them seem to do it quite often. In the world of desktop development, doesn't this create potential problems?

**Reply by thevortexcloud** [reply]
> So in such dynamically generated interactive buttons, the more common practice is "reflection binding"? 

It depends. If you know what the data type is ahead of time then you don't need to use a reflection binding at all.  If you have no idea (or are binding to `object`) then you can use a reflection binding and pray that it works. Ideally though you should avoid reflection bindings when possible. But if you use a reflection binding you also don't need to do any casting like I showed in my example either.

That said, what you are doing is odd (but not necessarily a bad thing depending on what the goal is). Typically each of your items would have their own view models with their own command declarations. You would then bind that instead of the parent view model, which would also avoid the cast since you have the correct strongly typed data context set for the template. But I have had to do similar things to what you have done myself many times.

> But in the official MusicStore example

I believe the MusicStore example actually predates compiled binding being enabled by default for new projects. I am not sure if it has been updated or not to account for that. In any case, yes: you should try to avoid reflection bindings when possible. You can change how `Binding` behaves to make it always use reflection/compiled bindings with an option in your project file. I personally though always explicitly use `CompiledBinding` and `ReflectionBinding` in my XAML to be more clear about what I am doing.

> doesn't this create potential problems?

It can, and does. This is how WPF works as there is no concept of compiled bindings at all there. It's fairly annoying to find these sorts of bugs there. 

The other advantage with compiled bindings is they work better with AOT.

**Reply by coinsoundsbetter** [reply]
I think I see what you mean. What you mean is that the ListBoxItem in the ListBox, for example, sometimes has its own ViewModel. So in the same scenario, they only need to bind to their own Command. Since the dynamic button in my example only needs to display a name represented by a string, I didn't create a separate ViewModel for it. Therefore, you can also simply handle it in this way (which is strange, but doesn't seem to be a big problem in the case of simple project size).

**Reply by thevortexcloud** [reply]
Pretty much. To be clear though, sometimes what you have done is the best way to do it as you may need some data/command in the parent view model. EG you may have a list of buttons to trigger a command on the parent, and then pass command parameter to the parent to change what the command does. I typically use this to have list of options with a delete button as the button needs to change what happens in the parent.

**Reply by coinsoundsbetter** [reply]
I see what you mean! Amazing!
